Source/DafnyCore/Auditor/Assumption.cs

```diff
@@ -0,0 +1,211 @@
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Dafny.Auditor;
+
+/// <summary>
+/// A set of assumption-related properties of a Dafny entity. Certain
+/// combinations of these properties classify the entity as containing
+/// an assumption.
+/// </summary>
+public struct AssumptionProperties {
```

Comment:
An alternative to what's there right now but perhaps compatible with what @keyboardDrummer is suggesting would be to have a separate traversal (not a set of instance methods on the nodes) that goes directly from the tree to a set of (issue, mitigation) pairs. That is, `GetAssumptionTags` could be called `GetAssumptionDescriptions` and return a list of `(string, string) ` pairs (or a record) directly.

I'd be quite happy to move in that direction. Early in the design I wanted to avoid doing many traversals over the whole tree, but the way things worked out that wouldn't happen even with the direct implementation.

Reply:
> And I think it's appropriate to encode this information in a separate struct than trying to embed it deeply in AST nodes.

Why? I don't see a downside to putting this information into the AST nodes. I would say that whatever is mandatory for Dafny, such as parsing, printing and resolution, should be in the AST nodes. I think this auditor is similar to warnings generated during resolution.

> which in effect has the same effect as having a struct that contains assumptions.

It's not the same, because it's easy to see what methods you can override on a class, but it's not easy to see every place where all ASTs nodes are handled by a switch.

Reply:
> Why? I don't see a downside to putting this information into the AST nodes. I would say that whatever is mandatory for Dafny, such as parsing, printing and resolution, should be in the AST nodes. I think this auditor is similar to warnings generated during resolution.

So you want to make the auditor mandatory? It looks like that's the single point of debate. If it's not mandatory, then it makes sense to keep it separate. If it's mandatory, then yes of course we should just put all its information into the AST warnings and get rid of the auditor itself.
I believe it's not mandatory as it could reduce productivity, as explained in another post.

Reply:
Having thought about this some more, I think I'm convinced that embedding the assumption descriptions in the AST _does_ make sense, in part because it'll make sure that there's a clear place to put any description of assumptions that may be present in any language construct as the language evolves. That's just a question of where assumption descriptions exist, though, not a question of when or how they should be reported.

I firmly believe that there are multiple use cases, each of which warrants different "warnings", or at least different output. The key long-term goal of the auditor, in my mind, is to give a complete view of the possible caveats to the verification of a particular program. Some of those caveats may be completely unavoidable (i.e., "we assume that Z3 v4.8.5 is correct", which is not currently in the auditor report but will ultimately be included), so they don't make sense as warnings during development but do make sense as part of a complete assurance summary.